components/Ink/src/private/MDCInkLayer.m

```diff
@@ -15,590 +15,195 @@
-      }
-    }];
-    [CATransaction commit];
-    return;
-  }
-
-  if (self.bounded) {
-    _foregroundOpacityAnim.values = @[ @1, @0 ];
-    _foregroundOpacityAnim.duration = kInkLayerForegroundBoundedOpacityExitDuration;
-
-    // Bounded ripples move slightly towards the center of the tap target. Unbounded ripples
-    // move to the center of the tap target.
-
-    CGFloat xOffset = self.targetFrame.origin.x - self.inkLayer.frame.origin.x;
-    CGFloat yOffset = self.targetFrame.origin.y - self.inkLayer.frame.origin.y;
-
-    CGPoint startPoint = CGPointMake(self.point.x + xOffset, self.point.y + yOffset);
-    CGPoint endPoint = MDCInkLayerRectGetCenter(self.targetFrame);
-    if (self.useCustomInkCenter) {
-      endPoint = self.customInkCenter;
-    }
-    endPoint = CGPointMake(endPoint.x + xOffset, endPoint.y + yOffset);
-    CGPoint centerOffsetPoint = MDCInkLayerInterpolatePoint(startPoint, endPoint, 0.3f);
-    UIBezierPath *movePath = [UIBezierPath bezierPath];
-    [movePath moveToPoint:startPoint];
-    [movePath addLineToPoint:centerOffsetPoint];
-
-    _foregroundPositionAnim =
-        [self positionAnimWithPath:movePath.CGPath
-                          duration:kInkLayerForegroundBoundedPositionExitDuration
-                    timingFunction:[self logDecelerateEasing]];
-    _foregroundScaleAnim.values = @[ @0, @1 ];
-    _foregroundScaleAnim.keyTimes = @[ @0, @1 ];
-    _foregroundScaleAnim.duration = kInkLayerForegroundBoundedRadiusExitDuration;
-  } else {
-    NSNumber *opacityVal = [self.presentationLayer valueForKeyPath:kInkLayerOpacity];
-    if (!opacityVal) {
-      opacityVal = [NSNumber numberWithFloat:0];
-    }
-    CGFloat adjustedDuration = kInkLayerForegroundBoundedPositionExitDuration;
-    CGFloat normOpacityVal = opacityVal.floatValue;
-    CGFloat opacityDuration = normOpacityVal / 3.f;
-    _foregroundOpacityAnim.values = @[ opacityVal, @0 ];
-    _foregroundOpacityAnim.duration = opacityDuration + adjustedDuration;
-
-    NSNumber *scaleVal = [self.presentationLayer valueForKeyPath:kInkLayerScale];
-    if (!scaleVal) {
-      scaleVal = [NSNumber numberWithFloat:0];
-    }
-    CGFloat unboundedDuration = (CGFloat)sqrt(((1.f - scaleVal.floatValue) * self.radius) /
-                                              (kInkLayerForegroundWaveTouchDownAcceleration +
-                                               kInkLayerForegroundWaveTouchUpAcceleration));
-    _foregroundPositionAnim.duration = unboundedDuration + adjustedDuration;
-    _foregroundScaleAnim.values = @[ scaleVal, @1 ];
-    _foregroundScaleAnim.duration = unboundedDuration + adjustedDuration;
-  }
-
-  _foregroundOpacityAnim.keyTimes = @[ @0, @1 ];
-  if (_foregroundOpacityAnim.duration < _foregroundScaleAnim.duration) {
-    _foregroundScaleAnim.delegate = self;
-  } else {
-    _foregroundOpacityAnim.delegate = self;
-  }
-
-  _foregroundOpacityAnim.timingFunction =
+- (void)setFrame:(CGRect)frame {
```

Comment:
Earlier setBounds question addressed elsewhere, but still exists here.
Do we need to handle setBounds?

Reply:
Setting this in setNeedsLayout to address both frame and bounds changes.